### PR TITLE
Refactor `UserToolchain`

### PR DESCRIPTION
**Pull request by neonichu**
My main goal here was to move the direct client of `ToolchainConfiguration` to use `UserToolchain` to ensure we are not calculating the host triple from scratch in several places. This required moving the toolchain configuration related types into `PackageModel` and cleaning up some deprecated APIs in the process.

I think a good next step here would be to actually make `ToolchainConfiguration` host all the directly toolchain related bits and have `UserToolchain` be the wrapper type which aggregates `ToolchainConfiguration`, `Destination` and adds the SDK/XCTest bits. This would give us a bit better separation of concerns. Potentially, we could also extract the SDK/XCTest bits into their own type and let `UserToolchain` just be the aggregator of the different sources of information. This should happen in subsequent PRs, though.

### CODE REVIEW

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
Note: all the older APIs I am removing here were deprecated in the 5.5/5.6 timeframe which I think makes it appropriate to remove them in 5.7

**Comment by neonichu** [comment]
Looks like this breaks `-[PackageLoadingTests.PackageDescription5_0LoadingTests testManifestLoaderEnvironment]`

File: Sources/PackageLoading/ManifestLoader.swift [hunk 1]
```diff
@@ -186,73 +184,6 @@ public final class ManifestLoader: ManifestLoaderProtocol {
         self.concurrencySemaphore = DispatchSemaphore(value: Concurrency.maxOperations)
     }
 
-    // deprecated 8/2021
```
Comment (tomerd) [inline Sources/PackageLoading/ManifestLoader.swift:189]: @neonichu could/should we do a separate PR to remove all the deprecated API, get that merged than do this PR on top?

File: Sources/PackageModel/Toolchain.swift [hunk 2]
```diff
@@ -11,11 +11,10 @@
 //===----------------------------------------------------------------------===//
 
 import TSCBasic
-import PackageModel
 
 public protocol Toolchain {
     /// Path of the `swiftc` compiler.
-    var swiftCompiler: AbsolutePath { get }
+    var swiftCompilerPath: AbsolutePath { get }
```
Comment (tomerd) [inline Sources/PackageModel/Toolchain.swift:17]: <3

File: Sources/PackageModel/UserToolchain.swift [hunk 3]
```diff
@@ -29,43 +26,23 @@ private let hostExecutableSuffix = ""
 public final class UserToolchain: Toolchain {
     public typealias SwiftCompilers = (compile: AbsolutePath, manifest: AbsolutePath)
 
-    /// The manifest resource provider.
-    public let configuration: ToolchainConfiguration
+    /// The toolchain configuration.
+    private let configuration: ToolchainConfiguration
```
Comment (tomerd) [inline Sources/PackageModel/UserToolchain.swift:30]: <3

**Comment by tomerd** [comment]
this is great!

**Comment by neonichu** [comment]
> Looks like this breaks -[PackageLoadingTests.PackageDescription5_0LoadingTests testManifestLoaderEnvironment]

The issue is that the `environment` argument to `UserToolchain` seems to be mostly not used.

File: Sources/PackageLoading/ManifestLoader.swift (hunk 1 above)
Comment (neonichu) [reply Sources/PackageLoading/ManifestLoader.swift:189]: Not sure, if you think it's valuable to split it out, I can definitely do that.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Review by tomerd — APPROVED** [review]

File: Sources/PackageLoading/ManifestLoader.swift (hunk 1 above)
Comment (tomerd) [reply Sources/PackageLoading/ManifestLoader.swift:189]: feels like the right order of things, also allows us to back out one without backing out the other

Comment (neonichu) [reply Sources/PackageLoading/ManifestLoader.swift:189]: https://github.com/apple/swift-package-manager/pull/4256

Comment (neonichu) [reply Sources/PackageLoading/ManifestLoader.swift:189]: That one removes all deprecated API that was deprecated 08/2021

**Comment by neonichu** [comment]
Will need to land #4256 first and then rebase this one.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Review by abertelrud — APPROVED** [review]
Thanks @neonichu!  This is great!

**Comment by compnerd** [comment]
This seems to have broken the windows build: https://ci-external.swift.org/job/oss-swift-windows-toolchain-x86_64-vs2019/331/console

**Comment by compnerd** [comment]
BTW, the breakage here is not visible due to the breakage introduced by the parent commit:

```
Sources\Commands\TestingSupport.swift:132:37: error: 'configuration' is inaccessible due to 'private' protection level
        if let location = toolchain.configuration.xctestPath {
                                    ^~~~~~~~~~~~~
PackageModel.UserToolchain (private):3:17: note: 'configuration' declared here
    private let configuration: PackageModel.ToolchainConfiguration
                ^
```
